src/load_balancer/cuda/replicated_load_balancer.cxx

```diff
@@ -0,0 +1,302 @@
+#include "replicated_load_balancer.hpp"
+#include <gauxc/util/div_ceil.hpp>
+#include <gauxc/util/cuda_util.hpp>
+
+#include <thrust/host_vector.h>
+#include <thrust/system/cuda/experimental/pinned_allocator.h>
+
+#include "load_balancer/cuda/cuda_collision_detection.hpp"
+
+using namespace GauXC::load_balancer::cuda;
+
+namespace GauXC {
+namespace detail {
+
+DeviceReplicatedLoadBalancer::DeviceReplicatedLoadBalancer( const DeviceReplicatedLoadBalancer& ) = default;
+DeviceReplicatedLoadBalancer::DeviceReplicatedLoadBalancer( DeviceReplicatedLoadBalancer&& ) noexcept = default;
+
+DeviceReplicatedLoadBalancer::~DeviceReplicatedLoadBalancer() noexcept = default;
+
+std::unique_ptr<LoadBalancerImpl> DeviceReplicatedLoadBalancer::clone() const {
+  return std::make_unique<DeviceReplicatedLoadBalancer>(*this);
+}
+
+
+std::vector<int32_t> inline copy_shell_list(
+  const size_t idx,
+  const std::vector<int32_t>& counts,
+  const thrust::host_vector<int32_t, thrust::cuda::experimental::pinned_allocator<int32_t>> &position_list
+) {
+  int32_t start = 0;
+  if ( idx != 0 ) start += counts[idx-1];
+  int32_t end = counts[idx];
+
+  std::vector<int32_t> shell_list(end - start);
+  std::copy(position_list.begin() + start, position_list.begin() + end, shell_list.begin());
+
+  return std::move( shell_list );
+}
+
+
+std::vector< XCTask > DeviceReplicatedLoadBalancer::create_local_tasks_() const  {
+
+  const int32_t n_deriv = 1;
+  const size_t atBatchSz = 256;
+
+  int32_t world_rank;
+  int32_t world_size;
+
+#ifdef GAUXC_ENABLE_MPI
+  MPI_Comm_rank( comm_, &world_rank );
+  MPI_Comm_size( comm_, &world_size );
+#else
+  world_rank = 0;
+  world_size = 1;
+#endif
+
+  std::vector< XCTask > local_work;
+  std::vector<size_t> global_workload( world_size, 0 );   
+
+  const auto natoms           = this->mol_->natoms();
+  const size_t nspheres       = (*this->basis_).size();
+  const size_t num_atom_batch = util::div_ceil(natoms, atBatchSz);
+  const size_t max_nbatches   = mg_->max_nbatches() * atBatchSz;
+  const size_t LD_bit         = util::div_ceil(nspheres, 32);
+
+  CollisionDetectionCudaData data;
+  cudaStream_t master_stream = 0;
+
+  std::vector< XCTask > temp_tasks;              temp_tasks.reserve( max_nbatches );
+  std::vector<std::array<double,3>> low_points;  low_points.reserve( max_nbatches );
```

Comment:
Yes—you're correct that using a SoA would improve the access patterns. We'd need to do some more profiling to determine how much of an impact it would make. Right now I think the collision detection kernel is fast enough, but it could make a difference elsewhere.

Reply:
OK, let's just keep things as-is and we can take a deeper dive to check some further optimizations in a future PR